From bef8518c4e9f40aabbe3bebf8f33eee8708ea376 Mon Sep 17 00:00:00 2001
From: Bai Ping <b51503@freescale.com>
Date: Fri, 19 Jun 2015 01:33:04 +0800
Subject: [PATCH] MLK-11117-02 ARM: imx6: update the clk switch code on imx6ul

Due to the clk tree info for AXI/AHB updating on i.MX6UL, the
busfreq enter/exit on i.MX6UL also need to be updated to align
to these changes. The AXI/AHB should be source from pll2_bus, and
the AHB clock divider value need to be set to 4 to make sure when
exiting from low bus mode, the AHB clock is also the original rate.

Signed-off-by: Bai Ping <b51503@freescale.com>
---
 arch/arm/mach-imx/busfreq-imx.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-imx/busfreq-imx.c b/arch/arm/mach-imx/busfreq-imx.c
index 5009267..2408e4d 100644
--- a/arch/arm/mach-imx/busfreq-imx.c
+++ b/arch/arm/mach-imx/busfreq-imx.c
@@ -268,12 +268,19 @@ static void exit_lpm_imx6_up(void)
 	 * lower ahb/ocram's freq first to avoid too high
 	 * freq during parent switch from OSC to pll3.
 	 */
-	imx_clk_set_rate(ahb_clk, LPAPM_CLK / 3);
+	if (cpu_is_imx6ul())
+		imx_clk_set_rate(ahb_clk, LPAPM_CLK / 4);
+	else
+		imx_clk_set_rate(ahb_clk, LPAPM_CLK / 3);
 	imx_clk_set_rate(ocram_clk, LPAPM_CLK / 2);
 	/* set periph_clk2 to pll3 */
 	imx_clk_set_parent(periph_clk2_sel, pll3);
+	/* set periph clk to from pll2_bus on i.MX6UL */
+	if (cpu_is_imx6ul())
+		imx_clk_set_parent(periph_pre_clk, pll2_bus);
 	/* set periph clk to from pll2_400 */
-	imx_clk_set_parent(periph_pre_clk, pll2_400);
+	else
+		imx_clk_set_parent(periph_pre_clk, pll2_400);
 	imx_clk_set_parent(periph_clk, periph_pre_clk);
 
 	if (ddr_type == MMDC_MDMISC_DDR_TYPE_DDR3)
-- 
1.8.0

